streams/shared/src/main/scala/zio/stream/StreamEffect.scala

```diff
@@ -0,0 +1,275 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.stream
+
+import zio._
+
+private[stream] trait StreamEffect[+E, +A] extends ZStream[Any, E, A] { self =>
+
+  def processEffect: Managed[E, () => A]
+
+  override def collect[B](pf: PartialFunction[A, B]): StreamEffect[E, B] =
+    new StreamEffect[E, B] {
+      def process = StreamEffect.super.collect(pf).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            def pull(): B = pf.applyOrElse(it(), (_: A) => pull())
+            () => pull()
+          }
+        }
+    }
+
+  override def collectWhile[B](pred: PartialFunction[A, B]): StreamEffect[E, B] =
+    new StreamEffect[E, B] {
+      def process = StreamEffect.super.collectWhile(pred).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            var done = false
+
+            () => {
+              if (done) throw StreamEffect.End
+              else pred.applyOrElse(it(), (_: A) => { done = true; throw StreamEffect.End })
+            }
+          }
+        }
+    }
+
+  override def dropWhile(pred: A => Boolean): StreamEffect[E, A] =
+    new StreamEffect[E, A] {
+      def process = StreamEffect.super.dropWhile(pred).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            var drop = true
+
+            def pull(): A = {
+              val a = it()
+              if (!drop) a
+              else if (!pred(a)) {
+                drop = false
+                a
+              } else pull()
+            }
+
+            () => pull()
+          }
+        }
+    }
+
+  override def filter(pred: A => Boolean): StreamEffect[E, A] =
+    new StreamEffect[E, A] {
+      def process = StreamEffect.super.filter(pred).process
+
+      def processEffect =
+        self.processEffect.flatMap { it =>
+          Managed.effectTotal {
+            def pull(): A = {
+              val a = it()
+              if (pred(a)) a else pull()
+            }
+
+            () => pull()
+          }
+        }
+    }
+
+  final def foldEffect[S](s: S)(cont: S => Boolean)(f: (S, A) => S): Managed[E, S] =
```

Comment:
`effectFold`, perhaps? Any method that captures a side-effect must be prefixed by `effect`. Also in this case, it's not clear where the effect is—`cont`, `f`, or both?

Is this really needed? I think it will be confusing for users. Fine to use internally of course.

@iravid Any other name ideas?

Reply:
The method is not exported, visible only in `zio.stream` package.

Reply:
We could call this `foldLeftManaged`, and also add it to `ZStream` itself. I'll add a ticket.

Reply:
Not sure `foldLeftManaged` is appropriate. `foldLeft` doesn't have a continuation function.

I'm thinking more of dividing the folds in effectful and non-effectful, depending on the state transformation function, but I hate the name `foldMManaged` and others...

Reply:
Dug up the old StreamPure, this used to be called `foldLazyPure`. @iravid